torch/distributed/_tensor/api.py

```diff
@@ -59,13 +60,24 @@
 #
 class _ToTorchTensor(torch.autograd.Function):
     @staticmethod
-    def forward(ctx, input: "DTensor", grad_placements: Optional[Sequence[Placement]]):  # type: ignore[override]
+    def forward(  # type: ignore[override]
+        ctx,
+        input: "DTensor",
+        grad_placements: Optional[Sequence[Placement]],
+        async_output: bool,
+    ):
         ctx.dtensor_spec = input._spec
         ctx.grad_placements = grad_placements
+        local_tensor = input._local_tensor
+        if not async_output and isinstance(local_tensor, funcol.AsyncCollectiveTensor):
+            # synchronously wait for any pending collectives to get the result tensor
+            local_tensor = local_tensor.trigger_wait()
+            local_tensor = local_tensor.elem  # type: ignore[attr-defined]
```

Comment:
I don't have a strong opinion on this, but this might bring some perf regression because for every `.to_local` we are waiting. But if we think this is the defined or expected behavior, then that's fine.